docs/source/kinds_of_types.rst

```diff
@@ -188,6 +188,130 @@ using bidirectional type inference:
 If you want to give the argument or return value types explicitly, use
 an ordinary, perhaps nested function definition.
 
+Extended Callable types
+***********************
+
+As an experimental mypy extension, you can specify ``Callable`` types
+that support keyword arguments, optional arguments, and more.  The
+"argument" part of the ``Callable`` specification (the part specifying
+the types of the arguments) can be a list of "argument specifiers"
+instead of a list of types.  This allows one to more closely emulate
+the full range of possibilities given by the ``def`` statement in
+Python.
+
+As an example, here's a complicated function definition and the
+corresponding ``Callable``:
+
+.. code-block:: python
+
+   from typing import Callable
+   from mypy_extensions import (Arg, DefaultArg, NamedArg,
+                                DefaultNamedArg, VarArg, KwArg)
+
+   def func(__a: int,  # This convention is for nameless arguments
+            b: int,
+            c: int = 0,
+            *args: int,
+            d: int,
+            e: int = 0,
+            **kwargs: int) -> int:
+       ...
+
+   F = Callable[[Arg(int),  # No argument name given
+                 Arg(int, 'b'),
+                 DefaultArg(int, 'c'),
+                 VarArg(int),
+                 NamedArg(int, 'd'),
```

Comment:
That discussion can go in https://github.com/python/typing/issues/264, which itself has a few naming differences from Naomi's PRs (#2426 and https://github.com/python/typeshed/issues/684). Or it can go into Naomi's PEP once she's written it.